internal/contentmanager/source/dynamicsource.go

```diff
@@ -0,0 +1,195 @@
+	}
+}
+
+// dynamicInformerSource is an implementation of the
+// ReaderCloserSyncingSource interface. It is used
+// to create an informer, using the provided dynamic informer
+// factory, for the configured GroupVersionResource.
+//
+// The informer is configured with a WatchEventErrorHandler that
+// stops the informer, and if it had previously synced successfully
+// it attempts to requeue the provided Owner for reconciliation and
+// calls the provided OnWatchError function.
+type dynamicInformerSource struct {
+	cfg            DynamicSourceConfig
+	informerCancel context.CancelFunc
+	informerCtx    context.Context
+	startedChan    chan struct{}
+	syncedChan     chan struct{}
+	erroredChan    chan struct{}
+	errOnce        sync.Once
+	err            error
+}
+
+func (dis *dynamicInformerSource) String() string {
+	return fmt.Sprintf("contentmanager.DynamicInformerSource - GVR: %s, Owner Type: %T, Owner Name: %s", dis.cfg.GVR, dis.cfg.Owner, dis.cfg.Owner.GetName())
+}
+
+func (dis *dynamicInformerSource) Start(ctx context.Context, q workqueue.TypedRateLimitingInterface[reconcile.Request]) error {
+	// Close the startedChan to signal that this
+	// source has been started. Subsequent calls
+	// to Start will attempt to close a closed channel
+	// and panic.
+	close(dis.startedChan)
+
+	dis.informerCtx, dis.informerCancel = context.WithCancel(ctx)
+	gInf := dis.cfg.DynamicInformerFactory.ForResource(dis.cfg.GVR)
+	eventHandler := source.NewEventHandler(dis.informerCtx, q, dis.cfg.Handler, dis.cfg.Predicates)
+
+	// If we encounter an error during the watch we will:
+	// - Capture the error
+	// - cancel the informer
+	// requeuing of the ClusterExtension should happen by the
+	// WaitForSync function returning an error
+	// Only if we have successfully synced in the past should we
+	// requeue the ClusterExtension
+	sharedIndexInf := gInf.Informer()
+	err := sharedIndexInf.SetWatchErrorHandler(func(r *cgocache.Reflector, err error) {
+		dis.errOnce.Do(func() {
+			dis.err = err
+			close(dis.erroredChan)
+		})
+
+		if dis.hasSynced() {
+			// We won't be able to update the ClusterExtension status
+			// conditions so instead force a requeue if we
+			// have previously synced and then errored
+			q.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: dis.cfg.Owner.GetName(),
+				},
+			})
+			dis.cfg.OnPostSyncError(dis.informerCtx)
+		}
+		dis.informerCancel()
+		cgocache.DefaultWatchErrorHandler(r, err)
+	})
+	if err != nil {
+		return fmt.Errorf("setting WatchErrorHandler: %w", err)
+	}
+
+	_, err = sharedIndexInf.AddEventHandler(eventHandler.HandlerFuncs())
+	if err != nil {
+		return fmt.Errorf("adding event handler: %w", err)
+	}
+
+	go sharedIndexInf.Run(dis.informerCtx.Done())
+
+	go func() {
+		syncOnce := sync.OnceFunc(func() {
+			dis.syncedChan <- struct{}{}
+			close(dis.syncedChan)
+		})
+
+		_ = wait.PollUntilContextCancel(dis.informerCtx, time.Second, true, func(ctx context.Context) (bool, error) {
```

Comment:
Nit: since we aren't using the context in the callback function.
```suggestion
		_ = wait.PollUntilContextCancel(dis.informerCtx, time.Second, true, func(_ context.Context) (bool, error) {
```

Reply:
done in https://github.com/operator-framework/operator-controller/pull/1119/commits/f6f277035a8331898fd304504b30660905797406